### PR TITLE
2.0.0: FIPS 202 SHA-3 standard hash algorithms and SHAVS test vectors

### LINKED ISSUE #14: Keccak != SHA-3
After the last update in 4/2014 SHA-3 is not exactly Keccak anymore. NIST added padding of 01.
So SHA3(256, "") != Keccak(256, "")
Please replace the name SHA-3 with Keccak to avoid confusion or make SHA3 compatible according to the actual NIST text vectors.
Thanx.

### PR DESCRIPTION
**Pull request by steakknife**
Closes #14

### CODE REVIEW

**Comment by steakknife** [comment]
Ready to merge and release.

cc: @FooBarWidget

**Comment by steakknife** [comment]
**TL;DR** Keccak reference implementation -> FIPS 202 patch 

```diff
--- KeccakSponge.c.orig	2016-03-21 00:20:19.000000000 -0700
+++ KeccakSponge.c	2016-03-20 11:07:20.000000000 -0700
@@ -200,13 +200,13 @@
 {
     // Note: the bits are numbered from 0=LSB to 7=MSB
     if (state->bitsInQueue + 1 == state->rate) {
-        state->dataQueue[state->bitsInQueue/8 ] |= 1 << (state->bitsInQueue % 8);
+        state->dataQueue[state->bitsInQueue/8 ] |= 6 << (state->bitsInQueue % 8);
         AbsorbQueue(state);
         memset(state->dataQueue, 0, state->rate/8);
     }
     else {
         memset(state->dataQueue + (state->bitsInQueue+7)/8, 0, state->rate/8 - (state->bitsInQueue+7)/8);
-        state->dataQueue[state->bitsInQueue/8 ] |= 1 << (state->bitsInQueue % 8);
+        state->dataQueue[state->bitsInQueue/8 ] |= 6 << (state->bitsInQueue % 8);
     }
     state->dataQueue[(state->rate-1)/8] |= 1 << ((state->rate-1) % 8);
     AbsorbQueue(state);

```

There are simpler, shorter implementations out there as well.

#### References
- [Keccak Round 3 finalist submission (zip)](http://csrc.nist.gov/groups/ST/hash/sha-3/Round3/documents/Keccak_FinalRnd.zip) from [SHA-3 Competition, Round 3](http://csrc.nist.gov/groups/ST/hash/sha-3/Round3/submissions_rnd3.html)
- [FIPS 202 Standard (pdf)](http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf) 
from [FIPS 202 Standard (landing)](http://csrc.nist.gov/groups/ST/hash/sha-3/fips202_standard_2015.html)
- [SHAVS test vectors for byte-length messages (zip)] (http://csrc.nist.gov/groups/STM/cavp/documents/sha3/sha-3bytetestvectors.zip)
from [SHAVS test vectors (landing)](http://csrc.nist.gov/groups/STM/cavp/secure-hashing.html#test-vectors)

**Comment by canterberry** [comment]
Sorry for the very long delay on this PR.

I completely agree with #14 that this library is a bit of a misnomer. There are consumers downstream that are using it intentionally to provide the Keccak variant. Even with a major version bump, if the library suddenly begins to produce a different hash, it could cause some trouble. To that end, I propose an iterative approach vs a rewrite.

Since the discussion on #14 is dated and closed, I'll create some follow-up issues to try and establish a plan for how to move forward, and we'll make this happen!

**Comment by canterberry** [comment]
Created #36 as a follow-up for a phased rollout of SHA-3.
